quality_of_service_demo/rclcpp/src/interactive_publisher.cpp

```diff
@@ -0,0 +1,169 @@
+
+#include "quality_of_service_demo/common_nodes.hpp"
+
+#include "./utils.hpp"
+
+using namespace std::chrono_literals;
+
+static constexpr char OPTION_HELP[] = "--help";
+static constexpr char OPTION_PUBLISH_DELAY[] = "--delay";
+static constexpr char OPTION_DEADLINE_PERIOD[] = "--deadline";
+static constexpr char OPTION_LIVELINESS_KIND[] = "--liveliness";
+static constexpr char OPTION_LEASE_DURATION[] = "--lease";
+
+void print_usage(const char * progname)
+{
+  std::cout << progname << " [OPTIONS]" << std::endl <<
+    std::endl << "Options when starting the demo:" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << OPTION_HELP <<
+    "print this help message" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << OPTION_PUBLISH_DELAY <<
+    "the amount of delay between publishing subsequent messages" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << OPTION_DEADLINE_PERIOD <<
+    "deadline period in seconds" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << OPTION_LIVELINESS_KIND <<
+    "liveliness kind" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << OPTION_LEASE_DURATION <<
+    "lease duration for liveliness in seconds" << std::endl <<
+    std::endl <<
+
+    "Commands when the demo is running:" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << 'n' <<
+    "manually assert the liveliness of the node" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << 'p' <<
+    "manually assert the liveliness of the publisher" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << 's' <<
+    "toggle start/stop of publishing messages" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << 'q' <<
+    "print the QoS settings of the publisher" << std::endl <<
+    std::left << std::setw(14) << std::setfill(' ') << 'x' <<
+    "exit the demo" << std::endl <<
+    std::endl;
+}
+
+class PublisherCommandHandler : public CommandGetter
+{
+public:
+  PublisherCommandHandler(rclcpp::executors::SingleThreadedExecutor * exec, Talker * publisher)
+  : exec_(exec), publisher_(publisher) {}
+
+  void handle_cmd(const char command) const override
+  {
+    const char cmd = tolower(command);
+
+    if (cmd == 'n') {
+      // manually assert liveliness of node
+      publisher_->assert_node_liveliness();
+    } else if (cmd == 'p') {
+      // manually assert liveliness of publisher
+      publisher_->assert_publisher_liveliness();
+    } else if (cmd == 's') {
+      // toggle publishing of messages
+      publisher_->toggle_publish();
+    } else if (cmd == 'q') {
+      // print the qos settings
+      publisher_->print_qos();
+    } else if (cmd == 'x') {
+      // signal program exit
+      exec_->cancel();
+      std::cout << "exiting the demo..." << std::endl;
+    }
+  }
+
+private:
+  rclcpp::executors::SingleThreadedExecutor * exec_;
+  Talker * publisher_;
```

Comment:
The `SingleThreadedExecutor` is a stack object, so it can't be in a smart pointer. But good point, i'll use a reference and `weak_ptr` for these two members respectively.

Reply:
@jacobperron Can you take a look again at the updated code?